File: laboratorios/lab03/codigo/nReinas.py
import numpy as np
from math import fabs

def imprimirTablero(tablero):
    for i in range(len(tablero)):
        print("(",i,", ",tablero[i],")")

def __nReinas(n, tablero, c):
    if c >= n:
        imprimirTablero(tablero)
        return True

    for i in range(n):
        tablero[c] = i
        if __estanSegurasHastaElI(tablero,c):
            if __nReinas(n, tablero, c+1):
                return True
            tablero[c] = 0
    return False

def nReinas(n):
    __nReinas(n, list(np.zeros(n, dtype=np.int)) ,0)

def __estanSegurasHastaElI(tablero, elI):
    for i in range(elI):
        for j in range(i+1,elI+1):
            if fabs(tablero[i] - tablero[j]) == fabs(i-j):
                return False
            elif tablero[i] == tablero[j]:
                return False
    return True

File: talleres/taller03/nReinas.java
/**
 *
 * @author sgilz
 */
public class nReinas {
    private static void imprimirTablero(int [] tablero){
        for (int i = 0; i<tablero.length; ++i)
            System.out.println("("+i+", "+tablero[i]+")");
    }

    private static boolean nReinas(int n, int[] tablero, int c){
        if (c >= n) {
            imprimirTablero(tablero);
            return true;
        } 

        for(int i=0; i<n; ++i){
            tablero[c]=i;
            if(estanSegurasHastaElI(tablero,c)){
                if(nReinas(n, tablero, c + 1)) return true;
                tablero[c]=0;
            }

        }
        return false;
    }
    public static void nReinas(int n){
        nReinas(n,new int[n],0);
    }
    private static boolean estanSegurasHastaElI(int[] tablero, int elI){
        for (int i = 0; i < elI; i++) {
            for (int j = i+1; j <= elI; j++) {
                if (Math.abs(tablero[i] - tablero[j]) == Math.abs(i-j)) return false;
                else if(tablero[i] == tablero[j]) return false;
            }
        }
        return true;
    }

    public static void main(String[] args){
        nReinas(4);
    }
}